src/backend/src/user_profile_model.rs

```diff
@@ -0,0 +1,163 @@
+            self.user_profile_map
+                .get(&(updated, user_principal))
+                .map(|p| p.0)
+        } else {
+            None
+        }
+    }
+
+    pub fn store_new(
+        &mut self,
+        user_principal: StoredPrincipal,
+        timestamp: Timestamp,
+        new_user: &StoredUserProfile,
+    ) {
+        if let Some(old_updated) = self.user_profile_updated_map.get(&user_principal) {
+            // Clean up old entries
+            self.user_profile_map.remove(&(old_updated, user_principal));
+        }
+        self.user_profile_updated_map
+            .insert(user_principal, timestamp);
+        self.user_profile_map
+            .insert((timestamp, user_principal), Candid(new_user.clone()));
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use candid::Principal;
+    use ic_stable_structures::{
+        memory_manager::{MemoryId, MemoryManager},
+        DefaultMemoryImpl,
+    };
+    use shared::types::{user_profile::StoredUserProfile, Timestamp};
+    use std::cell::RefCell;
+
+    const USER_1: &str = "xzg7k-thc6c-idntg-knmtz-2fbhh-utt3e-snqw6-5xph3-54pbp-7axl5-tae";
+    const USER_2: &str = "ufjdl-kewp5-bgfaq-d7k34-e5w62-nyad4-7r3s5-m2pt2-owqga-kcr5z-jae";
+
+    fn prepare_btrees() -> (UserProfileMap, UserProfileUpdatedMap) {
+        const USER_PROFILE_MEMORY_ID: MemoryId = MemoryId::new(3);
+        const USER_PROFILE_UPDATED_MEMORY_ID: MemoryId = MemoryId::new(4);
+        let memory = RefCell::new(MemoryManager::init(DefaultMemoryImpl::default()));
+        let user_profile_map = UserProfileMap::new(memory.borrow().get(USER_PROFILE_MEMORY_ID));
+        let user_profile_updated_map =
+            UserProfileUpdatedMap::new(memory.borrow().get(USER_PROFILE_UPDATED_MEMORY_ID));
+
+        (user_profile_map, user_profile_updated_map)
+    }
+
+    #[test]
+    fn test_find_by_principal_returns_profiles() {
+        let (mut user_profile_map, mut user_profile_updated_map) = prepare_btrees();
+
+        let user_principal =
+            StoredPrincipal(Principal::from_text(USER_1).expect("invalid user principal"));
+        let now: Timestamp = 12345667788223;
+        let user_profile = StoredUserProfile::from_timestamp(now);
+        user_profile_map.insert((now, user_principal), Candid(user_profile.clone()));
+        user_profile_updated_map.insert(user_principal, now);
+
+        let user_principal_2 =
+            StoredPrincipal(Principal::from_text(USER_2).expect("invalid user principal"));
+        let another_now: Timestamp = now + 400000000;
+        let user_profile_2 = StoredUserProfile::from_timestamp(another_now);
+        user_profile_map.insert(
+            (another_now, user_principal_2),
+            Candid(user_profile_2.clone()),
+        );
+        user_profile_updated_map.insert(user_principal_2, another_now);
+
+        let user_profile_model =
+            UserProfileModel::new(&mut user_profile_map, &mut user_profile_updated_map);
```

Comment:
I would insert these using the `user_profile_model.insert()` API but this works too!

Reply:
Why? In my mind it doesn't make sense to have a user_profile_model withuot the btrees. That's why I set them in the initialization.